### DISCUSSION TITLE
Is there a way to get all values generated for a pixel by the samples per pixel argument?

### DISCUSSION THREAD

**Discussion by harshana95** [mitsuba-renderer/mitsuba3#263] [Q&A] [answered]
My understanding is that multiple samples are generated for each pixel (by spp argument) and aggregated at the end. I want to store all these values in an array without reducing them to the final value. Can you please tell me the way to do it?

I tried to create a custom Film, but the issue in https://github.com/mitsuba-renderer/mitsuba3/issues/166#issue-1330095133 arises.

**Comment by njroussel** [comment, marked as answer]
There was a missing constructor in the trampoline bindings for `Film` plugins. It will be fixed in the next release, it's on `master` now: https://github.com/mitsuba-renderer/mitsuba3/commit/69f39ce93e13ce1cac598c3d2bac3af5390c12bd. This should fix the `missing constructor` error you were seeing.

Your understanding is correct, samples are accumulated/aggregated. The [box filter](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_rfilters.html#box-filter-box) will guarantee that your samples are not weighted across multiple (neighboring) pixels.
The proper implementation of what you're trying to achieve would indeed be a `Film`. This would also maybe require a derived `ImageBlock` class since that is where the accumulation actually happens. But this would require a lot of changes in the C++ part of the code.

A easier implementation would be to ignore the existing `mi.render` and `integrator.render` pipelines, but rather to directly use `integrator.sample()`. You can take a look at this tutorial: https://mitsuba.readthedocs.io/en/latest/src/rendering/scripting_renderer.html
This should give you an idea of how to spawn the initial rays and then you can load whichever `Integrator` you desire and use its `sample()` method instead of writing out the rendering loop manually as it is done in the tutorial.

**Reply by harshana95** [reply]
Thank you. I will try to do that. BTW if i do multiple renders with spp=1 and different seed values, can I achieve the same result?

**Reply by njroussel** [reply]
Yes, absolutely.
The only difference will be in performance, I expect it to be slower.

**Reply by harshana95** [reply]
Thank you very much!
